panel/panel.py

```diff
@@ -169,10 +196,95 @@ def fill_panel(window, types=None, codes=None, linter=None, update=False):
 
                     line_msg = format_row(lineno, error_type, item)
                     to_render.append(line_msg)
+                    item["panel_lineno"] = to_render.current_lineno()
 
         to_render.append("\n")  # empty lines between views
 
-    if to_render:
-        panel.run_command('sublime_linter_update_panel', {'text': "\n".join(to_render).strip()})
+    rendered_text = to_render.render() if to_render else None
+    run_update_panel_cmd(panel, text=rendered_text)
+
+
+# logic for updating panel selection
+
+def get_closest_region_dict(dic, colno):
+    dics = [
+        d
+        for error_dict in dic.values() for d in error_dict
+        if d.get("panel_lineno")
+        # if d["start"] <= colno <= d["end"]  # problematic line
+    ]
+    if not dics:
+        return
+    return min(dics, key=lambda x: abs(x["start"] - colno))
+
+
+def get_next_lineno(num, interval):
+    interval = set(interval)
+    interval.discard(num)
+    interval = list(interval)
+    interval.sort()
+
+    if num < interval[0] or interval[-1] < num:
+        return interval[0]
     else:
-        panel.run_command('sublime_linter_update_panel', {'text': "No lint errors.", 'clear_sel': True})
+        i = bisect.bisect_right(interval, num)
+        neighbours = interval[i - 1:i + 1]
+        return neighbours[1]
+
+
+def change_selection(panel_lineno, full_line=False, window=None):
+    panel = get_panel(window or sublime.active_window())
+    if not panel:
+        return
+
+    region = sublime.Region(panel.text_point(panel_lineno - 1, -1))
+    if full_line:
+        region = panel.line(region)
+
+    selection = panel.sel()
+    selection.clear()
+    selection.add(region)
+
+    # scroll selection into panel
+    if not panel.visible_region().contains(region):
+        panel.show(region)
+
+    # simulate scrolling to enforce rerendering of panel,
+    # otherwise selection is not updated (ST core bug)
+    panel.run_command("scroll_lines")
+
+
+def update_panel_selection(vid, lineno=None, colno=None, window=None):
+    if lineno < 0:
+        lineno = None
+    if colno <= 0:
+        colno = None
+
+    full_line = False
+    view_dict = persist.errors.get_view_dict(vid)
+    if not view_dict or util.is_none_or_zero(view_dict["we_count_view"]):
+        return
+
+    line_dicts = view_dict["line_dicts"]
+
+    if lineno:
+        if lineno in line_dicts:
+            full_line = True
+        else:
+            lineno = get_next_lineno(lineno, line_dicts)
+
+    lineno = 0 if lineno is None else lineno
+
+    line_dict = line_dicts[lineno]
+    region_dict = get_closest_region_dict(line_dict, colno or 0)
+
+    if not region_dict:
+        return
+
+    if full_line and colno is not None:
+        full_line = region_dict["start"] <= colno <= region_dict["end"]
+
+    panel_lineno = region_dict.get("panel_lineno")
+
+    if panel_lineno:
```

Comment:
The same. Is this really a truthy test or a `None` test?

Reply:
In other words: It is important to state if your data is (0, 0) based and when it is (1, 1) base. (1, 1) based is what we present the user. Bc something like this is known to err very easy.